src/Build/Construction/ProjectRootElement.cs

```diff
@@ -1677,19 +1677,33 @@ private void ReloadFrom(Func<bool, XmlDocumentWithLocation> documentProducer, bo
         {
             ThrowIfUnsavedChanges(throwIfUnsavedChanges);
 
-            XmlDocumentWithLocation document = documentProducer(preserveFormatting ?? PreserveFormatting);
-
-            // Reload should only mutate the state if there are no parse errors.
-            ThrowIfDocumentHasParsingErrors(document);
-
-            // Do not clear the string cache.
-            // Based on the assumption that Projects are reloaded repeatedly from their file with small increments,
-            // and thus most strings would get reused
-            //this.XmlDocument.ClearAnyCachedStrings();
+            var oldDocument = XmlDocument;
+            XmlDocumentWithLocation newDocument = documentProducer(preserveFormatting ?? PreserveFormatting);
+            try
+            {
+                // Reload should only mutate the state if there are no parse errors.
+                ThrowIfDocumentHasParsingErrors(newDocument);
 
-            RemoveAllChildren();
+                RemoveAllChildren();
 
-            ProjectParser.Parse(document, this);
+                ProjectParser.Parse(newDocument, this);
+            }
+            finally
+            {
+                // Whichever document didn't become this element's document must be removed from the string cache.
+                // We do it after the fact based on the assumption that Projects are reloaded repeatedly from their
+                // file with small increments, and thus most strings would get reused avoiding unnecessary churn in
+                // the string cache.
+                var currentDocument = XmlDocument;
+                if (!object.ReferenceEquals(currentDocument, oldDocument))
+                {
+                    oldDocument.ClearAnyCachedStrings();
```

Comment:
If newDocument and oldDocument represent the new and old version of the same file, wouldn't clearing the cache remove the overlap as well?

Reply:
This cache is implemented with reference counting. If two documents have exactly the same strings, these strings will have a ref-count of (at least) 2 and `ClearAnyCachedStrings` makes the ref-count go to 1 without actually removing any strings. Admittedly the name is a bit misleading.

That's what we see in the dumps. The number of strings in the cache is reasonable but they have huge ref-counts and for each reference there is an XML document kept alive by the cache, which the cache needs so it can later go "aha, you're releasing this document, let me look up all its strings and ref-- them".